libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp

```diff
@@ -124,9 +124,10 @@ void NavEKF2_core::ResetPosition(void)
             // clear the timeout flags and counters
             rngBcnTimeout = false;
             lastRngBcnPassTime_ms = imuSampleTime_ms;
-        } else if (imuSampleTime_ms - extNavDataDelayed.time_ms < 250) {
+        } else if (imuSampleTime_ms - extNavMeasTime_ms < 250) {
```

Comment:
I don't think we need to change this part, or at least in the EKF3 it also resets to the delayed position.

Reply:
I think that is because posResetSource is set to EXTNAV. So later statement is true
https://github.com/ArduPilot/ardupilot/blob/12c9578a6601042a2e34377254183eecbad76760/libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp#L129

Reply:
ok, but I wonder why the time on the extNavDataDelayed would be more than 250ms.  It should be arriving quite often so it should be very new.  I actually wonder if perhaps the issue is that it's newer than the imuSampleTime_ms so we end up with a huge number (which is actually a negative value).

Reply:
> but I wonder why the time on the extNavDataDelayed would be more than 250ms

I print extNavDataDelayed .time_ms to console. It is always 250~260 smaller than imuSampleTime_ms

>  It should be arriving quite often so it should be very new

I agree. It is really strange

Reply:
I change a way to fix this by adding a resetDataSource (stolen from EKF3)  to ResetPostion()
`void ResetPosition(resetDataSource resetSrc = DEFAULT)`

Reply:
> I wonder why the time on the extNavDataDelayed would be more than 250ms. 

We retrieve ext nav data based on  `imuDataDelayed.time_ms`
https://github.com/ArduPilot/ardupilot/blob/1db0feea59bfc7b531fff957d7eed4ff00b1dfb7/libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp#L403
I just test it, when `imuDataDelayed.time_ms` = 11693. a `extNavDataDelayed.time_ms` = 11691 is recalled
but then in `ResetPosition()`
https://github.com/ArduPilot/ardupilot/blob/1db0feea59bfc7b531fff957d7eed4ff00b1dfb7/libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp#L127
`imuSampleTime_ms` is 11944 at that time.

By the way, I think we use `gpsDataNew` in `ResetPosition()`
https://github.com/ArduPilot/ardupilot/blob/1db0feea59bfc7b531fff957d7eed4ff00b1dfb7/libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp#L105

Reply:
@chobitsfan, re the delayed vs new, I checked with Paul re which should be used and he suggested that we should use whichever is closest.  My personal feeling is that we should actually change the reset logic so that we request a position reset (by setting a flag or something) and then we trigger the position reset the next time a piece of data from the appropriate sensor is recalled. I.e. more like how it's done with external nav.  This will ensure that we reset most accurately I think.